aws/src/main/java/org/apache/iceberg/aws/LakeFormationAwsClientFactory.java

```diff
@@ -0,0 +1,201 @@
+
+public class LakeFormationAwsClientFactory
+    implements AwsClientFactory {
+  private Map<String, String> properties;
+  private Region region;
+  private String s3Endpoint;
+
+  public LakeFormationAwsClientFactory() {
+  }
+
+  @Override
+  public void initialize(Map<String, String> catalogProperties) {
+    this.properties = Maps.newHashMap(catalogProperties);
+    this.region = region(properties);
+    this.s3Endpoint = properties.get(AwsProperties.S3FILEIO_ENDPOINT);
+  }
+
+  @Override
+  public S3Client s3() {
+    return S3Client.builder()
+        .endpointOverride(URI.create(s3Endpoint))
+        .credentialsProvider(buildS3CredentialProvider())
+        .region(region)
+        .httpClientBuilder(UrlConnectionHttpClient.builder())
+        .build();
+  }
+
+  @Override
+  public GlueClient glue() {
+    return GlueClient.builder()
+        .httpClientBuilder(UrlConnectionHttpClient.builder())
+        .build();
+  }
+
+  @Override
+  public KmsClient kms() {
+    return KmsClient.builder()
+        .httpClientBuilder(UrlConnectionHttpClient.builder())
+        .build();
+  }
+
+  @Override
+  public DynamoDbClient dynamo() {
+    return DynamoDbClient.builder()
+        .httpClientBuilder(UrlConnectionHttpClient.builder())
+        .build();
+  }
+
+  private AwsCredentialsProvider buildS3CredentialProvider() {
+
+    GlueClient glue = glue();
+    String glueCatalogId = properties.get(AwsProperties.GLUE_CATALOG_ID);
+    String databaseName = properties.get(AwsProperties.LAKE_FORMATION_DB_NAME);
+    String tableName = properties.get(AwsProperties.LAKE_FORMATION_TABLE_NAME);
+
+    GetTableResponse response = glue.getTable(GetTableRequest.builder()
+        .catalogId(glueCatalogId)
+        .databaseName(databaseName)
+        .name(tableName)
+        .build());
+    boolean isRegisteredWithLakeFormation = response.table().isRegisteredWithLakeFormation() != null ?
+        response.table().isRegisteredWithLakeFormation() : Boolean.FALSE;
+
+    if (isRegisteredWithLakeFormation) {
+      return new LakeFormationCredentialsProvider(properties);
+    } else {
+      return LakeFormationCredentialsProvider.getAssumeRoleCredentialsProvider(properties);
+    }
+  }
+
+  private static Region region(Map<String, String> properties) {
+    String region = PropertyUtil.propertyAsString(properties, AwsProperties.CLIENT_ASSUME_ROLE_REGION, null);
+    if (region == null || region.isEmpty()) {
+      throw new RuntimeException("Region is empty");
+    }
+    return Region.of(region);
+  }
+
+  static class LakeFormationCredentialsProvider implements AwsCredentialsProvider {
+    private LakeFormationClient client;
+    private String tableArn;
+
+    LakeFormationCredentialsProvider(Map<String, String> properties) {
+      AwsCredentialsProvider awsCredentialsProvider = getAssumeRoleCredentialsProvider(properties);
+      String region = PropertyUtil.propertyAsString(properties, AwsProperties.CLIENT_ASSUME_ROLE_REGION, null);
```

Comment:
if default is null, we should just use `properties.get` and then do `Preconditions.checkArgument` that region must not be null